src/features/map/MapVisualizer.tsx

```diff
@@ -55,45 +70,143 @@ function getStaticMapOptions(){
   };
 }
 
-export default function MapVisualizer({ map, interactable = false }: Props): JSX.Element {
-  const memoizedCreateGraph = useMemo(() => createGraph(map), [map]);
+export default function MapVisualizer({
+  map,
+  interactable = false,
+  select,
+  doubleClick,
+  onNodeMoved,
+  edgeCreation,
+}: Props): JSX.Element {
+  const [mapState, setMapState] = useState(createGraph(map));
   const additionalOptions = interactable ? {} : getStaticMapOptions();
+  const edgeCreationMode = edgeCreation?.value;
+  const setEdgeCreationMode = edgeCreation?.setter;
+
+  function handleDoubleClick(event: any) {
+    if (doubleClick) {
+      const newNodeData = doubleClick(
+        Math.round(event.pointer.canvas.x / DISTANCE_MULTIPLIER),
+        Math.round(event.pointer.canvas.y / DISTANCE_MULTIPLIER),
+      );
+
+      setMapState((graph: GraphData) => {
+        return {
+          nodes: [
+            ...graph.nodes,
+            createNode(
+              random(100, 999999),
+              newNodeData.name,
+              newNodeData.type,
+              newNodeData.position,
+            ),
+          ],
+          edges: [...graph.edges],
+        };
+      });
+    }
+  }
+
+  function handleDragEnd(event: any) {
+    if (event.nodes.length === 1) {
+      const nodeMovedId = event.nodes[0];
+      const nodeMoved = mapState.nodes.find((n) => n.id === nodeMovedId);
+      if (nodeMoved && onNodeMoved) {
+        // using position of mouse pointer, so some drags can be little bit off,
+        // but I couldn't find exact position where node had moved ;(
+        const newX = Math.round(event.pointer.canvas.x / DISTANCE_MULTIPLIER);
+        const newY = Math.round(event.pointer.canvas.y / DISTANCE_MULTIPLIER);
+
+        nodeMoved.x = newX * DISTANCE_MULTIPLIER;
+        nodeMoved.y = newY * DISTANCE_MULTIPLIER;
+
+        onNodeMoved(newX, newY, nodeMoved.label ?? '');
+      }
+    }
+  }
```

Comment:
a to tutaj na 100% musimy dla noda recznie ustawiac wspolrzedne? powinnismy chyba wlasnie jakos z tego przesunietego noda chyba odczytac, moze warto lognąc tego eventa i sprawdzic co tam siedzi

Reply:
Lognalem i serio nie znalazłem nic ciekawego, weź proszę sproboj też pogrzebać bo no chwilę grzebałem J serio nic sensownego

Reply:
sure

Reply:
dobra huj zostawmy tak

Reply:
Nie zapomnij i tak zaokrąglić wyniku, żeby jakieś miliony cyfr po przecinku nie wjebywax do jsona